gogen/internal/gotypes/ordered_map.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gotypes is a playground package for demonstrating types that are
+// used in the generated Go code.
+package gotypes
+
+import (
+	"errors"
+	"fmt"
+)
+
+// RoutingPolicy_PolicyDefinition is the parent for the policy statement, used
+// a demonstration of an ordered map (for an `ordered-by user` YANG list).
+type RoutingPolicy_PolicyDefinition struct {
+	// Statement is an ordered map of policy statements.
+	//
+	// Note that the design here is to not use a pointer so that the empty
+	// value is usable. This is unlike the regular unordered lists where
+	// the various helpers need to reside on the parent struct in order to
+	// avoid a nil pointer exception on the raw map type.
+	Statement RoutingPolicy_PolicyDefinition_Statement_Map
+}
+
+// RoutingPolicy_PolicyDefinition_Statement represents an ordered-map element.
+type RoutingPolicy_PolicyDefinition_Statement struct {
+	DummyActions []string
+	Name         *string
+}
+
+// RoutingPolicy_PolicyDefinition_Statement_Map is a candidate ordered-map
+// implementation.
+type RoutingPolicy_PolicyDefinition_Statement_Map struct {
+	// TODO: Add a mutex here and add race tests after implementing
+	// ygot.Equal and evaluating the thread-safety of ygot.
+	//mu       sync.RWmutex
```

Comment:
Actually I will hold off pending our discussion below, I'm still not convinced it's the right time to add this since the rest of ygot doesn't deal with thread-safety. If we do want thread-safety I think we would want another suite of tests. I also think this can be added in a separate PR.